### PR TITLE
[cmake] generate a mbedtls-config.h without private headers

### PR DESCRIPTION
**Pull request by Moandor-y**
Fixes https://github.com/openthread/openthread/issues/5072.

### CODE REVIEW

File: third_party/mbedtls/CMakeLists.txt [hunk 1]
```diff
@@ -26,15 +26,37 @@
 #  POSSIBILITY OF SUCH DAMAGE.
 #
 
-set(OT_MBEDTLS_DEFAULT_CONFIG_FILE \"${CMAKE_CURRENT_SOURCE_DIR}/mbedtls-config.h\")
+set(OT_MBEDTLS_DEFAULT_CONFIG_FILE \"${PROJECT_BINARY_DIR}/mbedtls-config.h\")
```
Comment (bukepo) [inline third_party/mbedtls/CMakeLists.txt:29]:
```suggestion
set(OT_MBEDTLS_DEFAULT_CONFIG_FILE \"${CMAKE_CURRENT_BINARY_DIR}/mbedtls-config.h\")
```

File: third_party/mbedtls/CMakeLists.txt [hunk 2]
```diff
@@ -26,15 +26,37 @@
...
 
 set(ENABLE_TESTING OFF CACHE BOOL "Disable mbedtls test" FORCE)
 set(ENABLE_PROGRAMS OFF CACHE BOOL "Disable mbetls program" FORCE)
 
+find_program(UNIFDEFALL_EXE unifdefall)
+
 add_subdirectory(repo)
 
+if(UNIFDEFALL_EXE)
+    add_custom_target(mbedtls-config.h
```
Comment (bukepo) [inline third_party/mbedtls/CMakeLists.txt:41]: install the target.

File: third_party/mbedtls/mbedtls-config.h [hunk 3]
```diff
@@ -26,7 +26,8 @@
  *  POSSIBILITY OF SUCH DAMAGE.
  */
 
-#ifndef MBEDTLS_CONFIG_H
+#ifndef \
```
Comment (bukepo) [inline third_party/mbedtls/mbedtls-config.h:29]: add comment here to explain why this is necessary?

File: third_party/mbedtls/CMakeLists.txt [hunk 4]
```diff
@@ -26,15 +26,37 @@
...
+        ${UNIFDEFALL_EXE}
+            "'-D$<JOIN:$<TARGET_PROPERTY:ot-config,INTERFACE_COMPILE_DEFINITIONS>,';'-D>'"
+            "-I$<JOIN:$<TARGET_PROPERTY:ot-config,INTERFACE_INCLUDE_DIRECTORIES>,;-I>"
+            "-I$<JOIN:${OT_PUBLIC_INCLUDES},;-I>"
+            "-I${CMAKE_CURRENT_SOURCE_DIR}/repo/include"
+            -xc
+            "${CMAKE_CURRENT_SOURCE_DIR}/mbedtls-config.h" |
+            sed '/openthread-core-config.h/d' >
```
[the remark this answers is not in the document]
Comment (Moandor-y) [reply third_party/mbedtls/CMakeLists.txt]: Done.

File: third_party/mbedtls/CMakeLists.txt (hunk 1 above)
Comment (Moandor-y) [reply third_party/mbedtls/CMakeLists.txt]: Done.

File: third_party/mbedtls/mbedtls-config.h (hunk 3 above)
Comment (Moandor-y) [reply third_party/mbedtls/mbedtls-config.h:30]: Done.

File: third_party/mbedtls/CMakeLists.txt [hunk 5]
```diff
@@ -26,15 +26,37 @@
...
 
+find_program(UNIFDEFALL_EXE unifdefall)
+find_program(SED_EXE sed)
+
 add_subdirectory(repo)
 
+if(UNIFDEFALL_EXE AND SED_EXE)
+    add_custom_target(mbedtls-config.h
+        ${UNIFDEFALL_EXE}
+            "'-D$<JOIN:$<TARGET_PROPERTY:ot-config,INTERFACE_COMPILE_DEFINITIONS>,';'-D>'"
```
Comment (jwhui) [inline third_party/mbedtls/CMakeLists.txt:44]: nit: are single quotes necessary here? I noticed that the lines below do not use single quotes.

Comment (Moandor-y) [reply third_party/mbedtls/CMakeLists.txt:44]: Yes, they are necessary here. This makes definitions like `-DOPENTHREAD_PROJECT_CORE_CONFIG_FILE="openthread-core-simulation-config.h"` passed correctly. Without this the double quotes will be interpreted by shell and eventually the macro will be expanded to a string without quotes.

**Pull request merged**